WebApp/autoreduce_webapp/utilities/pagination.py

```diff
@@ -0,0 +1,156 @@
+"""
+Custom Paginator for the web application and associated Pages that are used to model each
+individual page in the pagination.
+
+This works in a similar way to the standard Paginator for Django but gives the added benefit of
+easier display names, as well as setting a maximum number of pages to display in the pagination
+
+This is currently used in the instrument_summary.html page
+"""
+
+
+# pylint:disable=too-few-public-methods
+class PageLimitException(Exception):
+    """ Custom Exception class to catch when page record limit is reached"""
+    pass
+
+
+class CustomPaginator(object):
+    """
+    CustomPaginator to allow for more complex functionality to be used without doing heavy
+    lifting in the django template code
+    """
+
+    # pylint:disable=too-many-instance-attributes
+    def __init__(self, query_set, items_per_page, page_tolerance, current_page):
+        """
+        :param query_set: All data to show
+        :param items_per_page: Number of items per page
+        :param page_tolerance: Number of pages to show
+        """
+        self.query_set = query_set
+        self.items_per_page = int(items_per_page)
+        self.page_tolerance = int(page_tolerance)
+        self.current_page_index = int(current_page)
+        self.next_page_index = 0
+        self.has_next = False
+        self.has_previous = False
+        self.previous_page_index = 0
+        self.page_list = []
+        self.display_list = []
+        self._validate_current_page()
+        self._construct_pagination()
+        self._set_next_and_previous()
+        self.current_page = self.page_list[self.current_page_index - 1]
+        self._create_display_list()
+
+    def _validate_current_page(self):
+        """
+        Ensure that the current page specified is valid
+        Update the page to min/max if outside of expected range
+        """
+        page_count = len(self.query_set) / self.items_per_page
```

Comment:
There seems to be an off by one error somewhere around here that stops me from accessing the last page in a list. On my VM webapp, I set the page size to 10 and attempt to click the final run range in the list, the link just returns me to the penultimate run range.

Reply:
I believe this should be fixed with the changes I just added. I was relying too heavily on python integer division - now using float division and `math.ceil`.

Reply:
This is now fixed.